File: shared_resources/src/main/java/com/thewizrd/shared_resources/utils/WeatherException.java
package com.thewizrd.shared_resources.utils;

import android.content.Context;

import com.thewizrd.shared_resources.R;
import com.thewizrd.shared_resources.SimpleLibrary;

public class WeatherException extends Exception {
    private final WeatherUtils.ErrorStatus errorStatus;
    private Context context;

    public WeatherException(WeatherUtils.ErrorStatus errorStatus) {
        this.errorStatus = errorStatus;
        this.context = SimpleLibrary.getInstance().getApp().getAppContext();
    }

    public WeatherUtils.ErrorStatus getErrorStatus() {
        return errorStatus;
    }

    @Override
    public String getMessage() {
        String errorMsg;

        switch (errorStatus) {
            case NOWEATHER:
                errorMsg = context.getString(R.string.werror_noweather);
                break;
            case NETWORKERROR:
                errorMsg = context.getString(R.string.werror_networkerror);
                break;
            case INVALIDAPIKEY:
                errorMsg = context.getString(R.string.werror_invalidkey);
                break;
            case QUERYNOTFOUND:
                errorMsg = context.getString(R.string.werror_querynotfound);
                break;
            case UNKNOWN:
            default:
                errorMsg = context.getString(R.string.werror_unknown);
                break;
        }

        return errorMsg;
    }
}

File: shared_resources/src/main/java/com/thewizrd/shared_resources/exceptions/WeatherException.kt
package com.thewizrd.shared_resources.exceptions

import com.thewizrd.shared_resources.R
import com.thewizrd.shared_resources.sharedDeps

enum class ErrorStatus {
    UNKNOWN, SUCCESS, NOWEATHER, NETWORKERROR, INVALIDAPIKEY, QUERYNOTFOUND
}

class WeatherException : Exception {
    val errorStatus: ErrorStatus

    constructor(errorStatus: ErrorStatus) : super() {
        this.errorStatus = errorStatus
    }

    constructor(errorStatus: ErrorStatus, cause: Throwable) : super(cause) {
        this.errorStatus = errorStatus
    }

    override val message: String
        get() {
            return when (errorStatus) {
                ErrorStatus.NOWEATHER -> {
                    sharedDeps.context.getString(R.string.werror_noweather)
                }
                ErrorStatus.NETWORKERROR -> {
                    sharedDeps.context.getString(R.string.werror_networkerror)
                }
                ErrorStatus.INVALIDAPIKEY -> {
                    sharedDeps.context.getString(R.string.werror_invalidkey)
                }
                ErrorStatus.QUERYNOTFOUND -> {
                    sharedDeps.context.getString(R.string.werror_querynotfound)
                }
                else -> {
                    // ErrorStatus.UNKNOWN
                    sharedDeps.context.getString(R.string.werror_unknown)
                }
            }
        }
}